src/core/toPath.js

```diff
@@ -102,139 +102,70 @@ const allowingArrays = fn => arg => {
-
-      // Move index after the bracket
-      index = nextBracketIndex + 1
-    }
-
-    // If array notation is enabled
-    if (isArrayNotation) {
-      // Check if next character is a string quote
-      const { quoted, quote } = isQuoteChar(str, index)
-
-      // If array index is a quoted string
-      if (quoted) {
-        // Move index after the string quote
-        index++
-
-        // Look for the next unescaped matching string quote
-        let endQuoteIndex, quotedIndex = index
-        do {
-          endQuoteIndex = str.indexOf(quote, quotedIndex)
-          quotedIndex = endQuoteIndex + 1
-        } while (endQuoteIndex !== -1 && str.charAt(endQuoteIndex - 1) === '\\')
-
-        // If no end quote found, stop if end of str is reached, or continue to next iteration
-        if (endQuoteIndex === -1) {
-          if (index !== str.length) path.push(str.substring(index))
-          break
-        }
-
-        // Add the content of quotes to the path, unescaping escaped quotes
-        path.push(unescapeQuotes(str.substring(index, endQuoteIndex), quote))
-
-        // Move index after end quote
-        index = endQuoteIndex + 1
-
-        // If next character is a closing square bracket, move index after it
-        if (str.charAt(index) === ']') index++
-
-        // Stop if end of str has been reached
-        if (index === str.length) break
-
-        // If next character is a dot, move index after it (skip it)
-        if (str.charAt(index) === '.') index++
-
-      } else { // If array index is not a quoted string
-
-        // Look for the closing square bracket
-        const closingBracketIndex = str.indexOf(']', index)
-
-        // If no closing bracket found, stop if end of str is reached, or continue to next iteration
-        if (closingBracketIndex === -1) {
-          if (index !== str.length) path.push(str.substring(index))
-          break
-        }
-
-        // Fetch the content of brackets and move index after closing bracket
-        const arrayIndexValue = str.substring(index, closingBracketIndex)
-        index = closingBracketIndex + 1
-
-        // If next character is a dot, move index after it (skip it)
-        if (str.charAt(index) === '.') index++
-
-        // Shorthand: if array index is the whole slice add it to path
-        if (arrayIndexValue === ':') {
-          path.push([undefined, undefined])
-        } else {
-
-          // Look for a slice quote
-          const sliceDelimIndex = arrayIndexValue.indexOf(':')
-
-          // If no slice quote found
-          if (sliceDelimIndex === -1) {
-            // Parse array index as a number
-            const nArrayIndexValue = Number(arrayIndexValue)
-
-            // Add array index to path, either as a valid index (positive int), or as a string
-            path.push(isIndex(nArrayIndexValue) ? nArrayIndexValue : arrayIndexValue)
+  const path = stringToPath2(str)
+  return str[0] === '.' ? ['', ...path] : path
```

Comment:
Did you read lodash's stringToPath ?
They have the exact same test to add an empty segment in front of the path
https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L6754

Reply:
Ahah, no I did not, but it makes sense. The first dot is special because it is the only one for which the left hand part has not already been added to the result array. Once inside the recursive function there is no way to know if the code is reading at the beginning of the original string or not so its too late.